src/cargo/core/compiler/job_queue.rs

```diff
@@ -284,129 +616,33 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
         // successful and otherwise wait for pending work to finish if it failed
         // and then immediately return.
         let mut error = None;
-        let total = self.queue.len();
-        let mut finished = 0;
         loop {
-            // Dequeue as much work as we can, learning about everything
-            // possible that can run. Note that this is also the point where we
-            // start requesting job tokens. Each job after the first needs to
-            // request a token.
-            while let Some((unit, job)) = self.queue.dequeue() {
-                queue.push((unit, job));
-                if self.active.len() + queue.len() > 1 {
-                    jobserver_helper.request_token();
-                }
-            }
+            self.spawn_work_if_possible(cx, jobserver_helper, scope, error.is_some())?;
 
-            // Now that we've learned of all possible work that we can execute
-            // try to spawn it so long as we've got a jobserver token which says
-            // we're able to perform some parallel work.
-            while error.is_none() && self.active.len() < tokens.len() + 1 && !queue.is_empty() {
-                let (unit, job) = queue.remove(0);
-                self.run(&unit, job, cx, scope)?;
-            }
+            info!(
+                "tokens: {}, rustc_tokens: {}, waiting_rustcs: {}",
```

Comment:
Should this perhaps get printed just before blocking? It seems sort of weird to print out this information between spawning/granting?